### ISSUE TITLE
NaNs in the flat variance array contaminate the x1d combined error estimate

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#7647]
_Issue [JP-3250](https://jira.stsci.edu/browse/JP-3250) was created on JIRA by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe):_

From Help Desk ticket INC0189669 (link at the bottom for reference): 

The user noticed the FLUX_ERROR column in an x1d product was an array of NaNs. Howard pointed out that the cal files have all zero values in the FLAT_VARIANCE 2d array, which then propagates through to the x1d files created by the extract_1d step. It appears that the flat variance array is all zero because all of the ERR arrays in the current flat-field reference files for the NIRSpec fixed-slit mode (using the prism) are also zero-filled. 

So two things would need to be updated to address the user's issue:
 # Find and fix the fact that the NaN values in the flat variance array are contaminating the computation of the combined error estimate, which is composed of Poisson, read noise, and flat variances, so we could then at least get error estimates that represent the Poisson and read noise only. 
 # The NIRSpec instrument team would need to compute estimated variance data for the flats and store them in the flat-field reference files used by the calibration pipeline. This should probably be captured separately in a CRDS ticket. 

 

Help desk link: <https://stsci.service-now.com/nav_to.do?uri=%2Fincident.do%3Fsys_id%3D5d12d758477ea190289c4f32736d4364%26sysparm_stack%3Dincident_list.do%253fsysparm_query%253dactive%253dtrue%26sysparm_view%3DJWST%26sysparm_view_forced%3Dtrue>

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=647577#comment-647577):_

From NIRSpec: also answered in Help Desk ticket INC0189910.

**Comment by stscijgbot-jp** [comment]
_Comment by [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=648042#comment-648042):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) possibly very dumb question: for issue # 1 (flat variance contaminating computation of error estimate), that sounds like it could be a general issue (not just NIRSpec). Is that right? If so, I'll add the other instruments to this ticket.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=648298#comment-648298):_

[Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe)  Yes, that probably is a general issue that applies to all instruments. And according to [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) the problem may not actually be in the extract_1d step, but rather in the resample/resample_spec step, which is apparently creating resampled ERR arrays that are all zero/NaN for these cases.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=650489#comment-650489):_

Fixing/updating extract_1d to disregard NaN's when computing the quadrature sum for the FLUX_ERR data sounds like a reasonable approach, but [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) raises a fair question as to whether that's an appropriate thing to do, especially if only some pixels in the variance arrays are NaN. That would result in varying contributions from one wavelength bin to the next in the FLUX_ERR array and the user would have no way of knowing which contributions were included for a given bin. Perhaps we should solicit opinions from folks in DMS WG as to what they think about this approach [Alicia Canipe](https://jira.stsci.edu/secure/ViewProfile.jspa?name=acanipe)

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=650643#comment-650643):_

We had set the flat_variance arrays in the reference files to zero while debating how to handle the errors in the flats, not realizing this would be an issue. We will redeliver flats with non-zero variance arrays. Because we have three components to the flats that are interrelated, our plan was to keep the D- and S-flat variances at zero and put real uncertainties only in the F-flat files; I think this should work given how the combined var_flat is calculated, but can someone confirm?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=650651#comment-650651):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) Yes, I believe it should work OK to only have non-zero variances in the FFLAT. That's the way it used to be and we did not seem to have the issue downstream when that was the case. And looking at the code in the flat-field step where it combines the ERR arrays from DFLAT, FFLAT, and SFLAT to produce a combined flat-field ERR estimate (which then gets stored in the VAR_FLAT array of the science data), it uses simple summation of the err values from d_flat, f_flat, and s_flat and hence as long as at least one of them has non-zero contributions, we should get a non-zero VAR_FLAT array.  NaN's in the ref file ERR arrays, on the other hand, could be problematic. It looks like the current code is not robust enough to handle NaN's in the input ERRs.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=650669#comment-650669):_

Thanks, [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse). Unfortunately, I think there might be a problem with the FFLAT file format. We need to be able to put a 1D error array as part of the FAST_VARIATION table extension. However, it looks like that currently only has four parts -  slit name, nelem, wavelength vector, and "data" vector. I'm guessing that is baked into the data model? In which case, that would need to be changed before we could deliver new files.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=651553#comment-651553):_

Relevant to the above, I just opened a separate ticket to add an error array to the F-flat data model (https://jira.stsci.edu/browse/JP-3304).

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=654434#comment-654434):_

[Tony Keyes](https://jira.stsci.edu/secure/ViewProfile.jspa?name=keyes) [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) [Peter Zeidler](https://jira.stsci.edu/secure/ViewProfile.jspa?name=zeidler) [Kayli Glidic](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kglidic) The PR #7711 has been created, which includes a workaround for this issue, by modifying the way the extract_1d step computes the total ERR values from the constituent variance arrays. Right now it just blindly sums them all together, which means that if any of the components have NaN's in them, the ERR ends up with NaN's. The PR modifies that behavior such that if any of the variance arrays has even just one NaN value in it somewhere, it excludes that entire variance array from the computation of the combined ERR values. That puts a bandaid on the situation, such that we no longer have ERR arrays full of NaN, but the question is whether that's really the right thing to do or not. The issue is that there's no obvious way for a user to know that their ERR values were computed from only some of the variances and not all of them, so they don't know that the ERR's are not really representative of their data.

So perhaps it's better to just leave things the way they are right now, because at least when a user sees ERR values that are all NaN, they know that something's not right (they just may not understand exactly what isn't right). The ultimate solution of course is to finish the work being done to add ERR values to the NRS FFLAT ref file, so that we don't get into this situation in the first place.

Comments?

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=660473#comment-660473):_

In my opinion, the best approach would be a simple nansum across the variance components, rather than ignoring any component that has any NaNs at all.  There are valid scientific use cases for NaN values in error arrays, and they should be handled properly without invalidating the whole error array. 

Assigning meaningful error values should be handled in the reference files; communicating to users what they mean should be handled in documentation.  The software should just sensibly propagate the inputs it's given.

**Comment by stscijgbot-jp** [comment]
_Comment by [Peter Zeidler](https://jira.stsci.edu/secure/ViewProfile.jspa?name=zeidler) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=660495#comment-660495):_

I agree with [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke)

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=660591#comment-660591):_

Yes, I agree that making sure all pixels have sensible values is a reference file task.  I don't expect that there will be scattered NaN values in the reference files in science areas, but I imagine it might be useful for regions of the array to be set to NaN, if they are outside the usable area for a particular mode, for example.  If the software allows the error array to contain at least some NaNs, the science teams can make that decision for their own reference files.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=660592#comment-660592):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) that makes sense.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=661817#comment-661817):_

Can't we just set some small error value for all pixels until we get the actual errors?

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=661842#comment-661842):_

[Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) - Yes, that's possible, but there are three reasons why that's not the preferred solution for NIRSpec:
 # Reference file re-creation and delivery for all existing flats is a significant effort. We prefer to do it once, when the actual errors are known.
 # 0 or NaN for the error value is a more accurate scientific representation of the current situation than adding arbitrary small errors. The errors are currently unknown; assigning them a specific value is inaccurate.
 # Allowing NaN values in one error array to invalidate all the input error components is a software bug, and it should be fixed in the software. 0 and NaN values have valid scientific meaning in error arrays and should be propagated sensibly.

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=661857#comment-661857):_

Well, we can have a discussion on how both 0  and NaN error values has meaning. I withdraw my comment.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=661863#comment-661863):_

Unless we change the definition of the error array as currently generated, it is not a software bug - the error array as defined has three contributions. "Sensibly propagating" NaN values by removing them from the calculation fundamentally alters the definition of the generated array; if this is the desired outcome for all instrument teams, we can change the definition of the array. This puts the onus on the user noticing that entries in the spectral table with a NaN entry in a variance column means their error value will not be the quadrature sum of three contributions but of two (or less).

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=661923#comment-661923):_

Okay, software requirements change, then!  If this needs a cross-instrument decision, is there a way that can be done speedily to address this issue?

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=662048#comment-662048):_

following up here after an email exchange – with [Greg Sloan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gcsloan) now responsible for overseeing all spectroscopic aspects of the pipeline (while he continues transitioning to full CalWG lead) he will directly coordinate the rest of the discussion for this issue, including also the relevant spectroscopy mode experts from the other instruments, to ensure a path forward that works for NIRSpec (which may include evolving requirements, if that turns out to be necessary) and also maintaining consistency with pipeline processing for spectroscopy data on the other instruments.

**Comment by stscijgbot-jp** [comment]
_Comment by [Greg Sloan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gcsloan) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=663022#comment-663022):_

I am now fielding a helpdesk ticket that may be reporting a similar problem with x1d LRS data.  That's ticket INC0193278.  For the record, we may need to "fundamentally alter" the propagation of uncertainties to trap NaNs.  It wouldn't be that different from ignoring NaNs when calculating a median or mean of a set of data, except here the values are adding in quadrature.  The point of the uncertainties is to estimate the quality of the data, and they can't do that if they're NaNed.  I have always treated uncertainties as an approximation, and if we have 1 of 3 or 4 inputs NaNed, then we have to approximate the uncertainty with what we have left.  I would imagine that this point may require some discussion.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=668550#comment-668550):_

Adding a note that one of the specific avenues that this is causing a problem is through the pixel replacement step.  For the MIRI MRS at least, if pixel replacement is not used then everything works ok: NaN-valued ERR pixels are also NaN-valued fluxes and DO_NOT_USE DQ values.  If using the 'mingrad' pixel replacement everything also works ok, as new values are interpolated for both the SCI and ERR arrays and DQ values are flipped from DO_NOT_USE to being used.  If the 'fit_profile' method is used though, new values are interpolated for the SCI array and the DQ array is set to use them, but the ERR values are never changed from NaN.  This then propagates downstream.

I think this needs to be fixed in the pixel replacement step, as cube building is predicated on the notion that pixels with good-values fluxes and DQ should not have NaN errors.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=674643#comment-674643):_

Updated NRS FFLAT ref files delivered in context 1166 as part of B10.0 installation in Ops, so that VAR_FLAT arrays are now populated in science products (see JP-3304). Does that effectively close this ticket as well?

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=674648#comment-674648):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) - I think NaN values in error arrays still need to be handled better and more consistently throughout the pipeline.  We should no longer have the issue where an entire variance component is set to NaN, but there are sometimes scattered NaN values for other reasons, and it still might be useful to set errors for invalid regions to NaN without worrying that they will contaminate later propagation.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=681390#comment-681390):_

[Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) [Greg Sloan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gcsloan) [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) I think this ticket would be a good candidate for JPCT or CCT discussion, when such meetings begin or resume.

**Comment by stscijgbot-jp** [comment]
_Comment by [Anton Koekemoer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=koekemoe) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=681396#comment-681396):_

Thanks [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly). Following [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse)'s comment, I'd say the next step would be to first ensure that we have good consistency across the reference files of the various instruments, in terms of when NANs are / aren't used.

Once the reference files are confirmed to be consistent across the instruments in terms of their treatment of NANs, then we can return to the algorithm side in terms of how these pixels are handled in the pipeline code.

So I'm going to propose handing this to [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) to organize discussion of this in the JPCT meetings and ensure that there's convergence on that side, then this can come back to the JWST CCT side once we've confirmed that the reference files are consistent in this.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=681403#comment-681403):_

[Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) Sounds good.  We're still figuring out who all the instrument reps to the JP Coord Team will be, but I've added it to the agenda for the first meeting (tentatively next week, but TBD).

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=682733#comment-682733):_

Adding an example from a helpdesk ticket for discussion: [INC0196845](https://stsci.service-now.com/nav_to.do?uri=%2Fincident.do%3Fsys_id%3D0bb3e2a5477b7910289c4f32736d43b9%26sysparm_domain%3Dnull%26sysparm_domain_scope%3Dnull%26sysparm_record_list%3Dactive%3Dtrue%5Eassignment_group.nameSTARTSWITHNIRSpec%5Ecompany%3Dc3478b5edb013e0042685434ce961951%5EORwatch_listCONTAINS3f61c9d1db8be2000c0a5595ce961987%5EORcompany%3Dc3478b5edb013e0042685434ce961951%5EORcaller_id%3D3f61c9d1db8be2000c0a5595ce961987%5EORopened_by%3D3f61c9d1db8be2000c0a5595ce961987%5EORDERBYDESCopened_at%26sysparm_record_row%3D8%26sysparm_record_rows%3D556%26sysparm_view%3DJWST%26sysparm_view_forced%3Dtrue)

The data is from PID 3788; for example, file jw03788007001_03101_00002_nrs2_rate.fits, reduced with pipeline version 1.12.5.

The error arrays for the extracted spectra from this program have large regions that are set to NaN.  The root cause is that there are a number of pixels with slightly negative values in the rate image, since the rate is close to zero for a significant portion of the array.  These have Poisson variance set to zero, which is appropriate: the total error estimate should come from the read noise only for these pixels.  When the zero variance values pass through the resample step, they are set to NaN instead in the resampled Poisson variance component.  Then, when the final error is combined for the extracted data, the NaN values are not ignored, and they contaminate the final error estimate.

The workaround is to set NaN values in the s2d file to zero and re-extract the spectrum.  A better fix would be for the resample step to ensure that zero values in the variance components are handled as valid numerical values.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=689786#comment-689786):_

As an update, this ticket will be significantly affected by <https://jira.stsci.edu/browse/JP-3570> and <https://jira.stsci.edu/browse/JP-3569>

These should ensure that there are never NaN values in the resampled ERR or VAR arrays as zero-valued VAR_POISSON or VAR_FLAT should propagate as expected.  As such no workaround should be necessary when extracting spectra.

Setting to low priority for now, to be reassessed when the above tickets are done to see if it's actually fixed the issue reported here.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=696675#comment-696675):_

Adding a note that the example data here is indeed significantly improved by the draft PR described on <https://jira.stsci.edu/browse/JP-3569>

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=702483#comment-702483):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) I think this ticket should now be taken care of by virtue of the other tickets that have been dealt with lately (and the one outstanding on interpolating errors in pixel replacement), would you concur?  The original ticket about NRS slit data now looks ok, presumably your MOS test case should as well?

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=702488#comment-702488):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) - I think most of the known sources of NaNs in error arrays are now taken care of, yes. 

However, I think it would be more future-proof to also ignore NaN values in the sum over the errors for the extracted spectrum.  As this ticket's history shows, it's pretty easy to accidentally introduce NaNs to the error arrays, and it's less surprising to the user to ignore them than to invalidate the entire error sum.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=703603#comment-703603):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) Agreed, and now (or at least, once JP-3570 is finished) there should be no reason to be using such pixels.  Any objection to opening that as a separate ticket targeted more specifically at ignoring NaNs in extraction regions?

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=703606#comment-703606):_

Works for me!

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3250?focusedCommentId=703610#comment-703610):_

Closing this ticket as effectively completed via linked tickets.  Some related long-term work remains though, now tracked through the more specific tickets JP-3570 and JP-3655.
